### ISSUE TITLE
Output from `no_std` tests?

### ISSUE THREAD

**Issue by cbiffle** [rust-lang/miri#2597]
I'm looking into using Miri to test some `thumbv7em-none-eabihf` code from an operating system kernel. I was curious if there's any supported way to get output (e.g. stdout) from a `no_std` test right now, or to `abort`.

Approaches I've tried:

- If I can extract my code into a lib that isn't necessarily `no_std` and supports a platform where `std` is available (the test host, or if I want 32-bit, something like `armv7-none-linux-eabihf`), then I can run it with the test harness and get status/failure reporting.
- If I can construct my test as a `no_std`/`no_main` program using a `#[start]` fn, I can return a handful of bits of information, but only in the exit code from `start` -- in particular, it's not clear that I can `abort` from a deep place in the callstack, since `abort` isn't in `core`.

Since Miri appears to detect and emulate certain I/O routines on std platforms, I wonder if it's possible to call into these from a `no_std` environment? This would give a test-only facility similar to "semihosting" on ARM embedded targets, where the test harness could call magic functions. (Concretely, the two magic functions I want are: write byte sequence to stdout, and abort; this would let me implement a `panic_handler` that does something other than hang the process, which appears to be the only option in a test right now.)

From web searches, it looks like I might be the second user of `MIRI_NO_STD` after your test suite, so I could not find any worked examples to crib from. :-)

Thanks in advance!

**Comment by bjorn3** [comment]
The OS interface miri emulates depends on the target_os key in the target specification, which for your target is "none". As such miri doesn't emulate any OS. It should be possible to modify miri to add another OS interface for your target. As for aborting you can call the unstable core::intrinsics::abort() function. It is available on every target.

**Comment by saethlin** [comment]
Can you link to any documentation for these magic functions you use for semihosting? It sounds like maybe we could shim those to do... something?

**Comment by cbiffle** [comment]
> Can you link to any documentation for these magic functions you use for semihosting? It sounds like maybe we could shim those to do... something?

You probably don't want to emulate the ARM semihosting interface _per se_ because it's implemented by particular use of a `bkpt` machine instruction -- below the level where you're emulating, and ARM-specific. That said, [here are some decent docs on the interface](https://wiki.segger.com/Semihosting), and [here's the canonical Rust interface](https://docs.rs/cortex-m-semihosting/latest/cortex_m_semihosting/).

It's a broader interface than what I'd need for testing -- in particular it allows for input and filesystem access on the host. Though I guess such things might be useful.

> The OS interface miri emulates depends on the target_os key in the target specification, which for your target is "none". As such miri doesn't emulate any OS. It should be possible to modify miri to add another OS interface for your target.

As an operating system kernel, `none` is the typical OS component of the target triple. I suppose we could alter the build system to build for a non-`none` platform and use Miri's emulation of that. Probably the closest thing to my original suggestion would be to make up a _new_ `target_os`, such as `miri`, and provide a basic interface there. (Are either of these what you were suggesting, or did I misunderstand?)

I'd argue for providing an emulated interface for `target_os = "none"`, since this will be a _very_ common need in `no_std` contexts. Currently code like `tests/pass/no_std.rs` that is executed on a `target_os = "none"` triple only has the ability to return or hang, as far as I can tell.

> As for aborting you can call the unstable core::intrinsics::abort() function. It is available on every target.

Ah, I hadn't noticed that. Under Miri, calling that causes the program to exit with status code 0, so it doesn't enable me to report test success/failure. But that _does_ let me at least write a `panic_handler` that does something other than hang, so, that's nice.

**Comment by saethlin** [comment]
It looks to me like the writing to host stdout/stderr is pretty simple, I can probably whip up an implementation of it in Miri. Do you have an example program I can try out a prototype on?

**Comment by oli-obk** [comment]
One thing you could try is to use

```rust
extern "C" {
    fn abort() -> !;
    fn putc(c: u8);
}
```

I think miri will just interpret those even on unsupported targets

**Comment by saethlin** [comment]
I would indeed quite like to have a magic Miri library that works like this, but I don't know how to implement that.

The shim infrastructure checks targets, as far as I can tell:
```
error: unsupported operation: the target `none` is not supported
  --> src/main.rs:20:9
   |
20 |         printf("hi\0".as_bytes().as_ptr().cast());
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the target `none` is not supported
   |
```
which is probably the correct thing to do.

> I don't think emulating that is the right choice.

Can you expand on why? Personally, I would like to run code under Miri without modification to the code. And if there is hackery that needs to be done in order to execute code under Miri, I'd rather it be done in one place inside this codebase than spread across the ecosystem (or done in one-off things when debugging). But I just have no idea if these things apply to embedded programming. I just don't know what that's like.

Last night I whipped up a _very_ crude implementation of the semihosting APIs for printing in Miri. Maybe I'll have time to touch it up today https://github.com/rust-lang/miri/compare/master...saethlin:miri:cortex-m-semihosting

**Comment by oli-obk** [comment]
> I would indeed quite like to have a magic Miri library that works like this, but I don't know how to implement that. (Best I can do is spooky magic symbols you have to declare yourself, or sticking definitions for those in some crate you can use. I like neither.)

We could inject a module into the sysroot that all programs compiled under miri then have access to

**Comment by cbiffle** [comment]
> I think miri will just interpret those even on unsupported targets

Heheheh I will try this. Probably not the ideal long term solution but who knows!

>  > I don't think emulating that is the right choice.
> Can you expand on why? Personally, I would like to run code under Miri without modification to the code.

Certainly! I would also like to run code under Miri without modification. The cortex-m-semihosting API is

- ARM-specific -- my software runs on a variety of architectures
- Merely one of several ways of getting debug printfs etc out of even ARM-specific software -- we don't use it in production code because it _hits breakpoints when used_ and is thus incompatible with realtime applications. There's the semihosting mechanism, there's the [ITM](https://developer.arm.com/documentation/ddi0337/e/BABCCDFD) peripheral that's accessed by memory mapped I/O, there are vendor-specific trace peripherals or UARTs, ...etc.

In general I expect _essentially all_ embedded code to require modification to run on Miri, either because its preferred I/O facility is not present, or because it spends a lot of its time doing stuff like MMIO to peripherals that are not emulated by Miri (and rightfully so, QEMU has a hard time getting these peripherals right already). Us embedded folk are already pretty used to programming against thin abstraction layers that get implemented for different targets, because the targets are so wildly different -- so if I need to implement one against a Miri-specific API that's no big deal compared to doing it for ST ARM Cortex with their peripherals, NXP ARM Cortex with _their_ peripherals, ARM A-class, AArch64, AVR, RISC-V, .....

Does that make sense?

**Comment by RalfJung** [comment]
We already have [magic Miri-provided extern functions](https://github.com/rust-lang/miri/#miri-extern-functions). We can just add another one for printing to stdout (and probably another one for stderr).

`abort` already exists as an intrinsic, so... I guess we could also have it as a Miri extern function but it seems less necessary.

**Comment by cbiffle** [comment]
@RalfJung that sounds pretty reasonable. It looks like this would be a matter of defining the function in `src/shims/foreign_items.rs`? Would it need to do anything fancier than write to interpreter `stdout`/`stderr`? (I suppose I could check the various fs shims for examples.)

**Comment by RalfJung** [comment]
> Would it need to do anything fancier than write to interpreter stdout/stderr?

Not really, no.

**Comment by cbiffle** [comment]
So, something like this then? https://github.com/cbiffle/miri/commit/48f09537307d2648596dae7419de220df7647702

(I'd extend it to stderr before submitting a PR.)

**Comment by RalfJung** [comment]
Yeah that looks good, just please move the `use` up to the file level.

**Comment by cbiffle** [comment]
Fixed in #2606 .
